# 09_production_nlp: Population Stability Index (PSI) Drift Monitoring

This notebook validates production drift metrics. It implements Population Stability Index (PSI) and Wasserstein Distance using NumPy and SciPy, verifying them against our hand calculations.

## 1. Population Stability Index (PSI) Calculation

In [1]:
import numpy as np

# Proportions matching our hand calculations
expected = np.array([0.6, 0.3, 0.1])
actual = np.array([0.5, 0.3, 0.2])

def calculate_psi(p, q):
    p = np.clip(p, 1e-15, 1.0)
    q = np.clip(q, 1e-15, 1.0)
    return np.sum((p - q) * np.log(p / q))

psi_val = calculate_psi(actual, expected)
print(f"Calculated PSI Score: {psi_val:.4f}")

# Verify exact match with hand calculation score of 0.0875
np.testing.assert_almost_equal(psi_val, 0.0875, decimal=4)

Calculated PSI Score: 0.0875


### Output Analysis: PSI Score
The computed PSI score is exactly `0.0875`. Since the PSI score is less than the standard stability threshold of `0.10`, the difference between expected (training) and actual (production) distribution represents no significant shift, indicating a stable production pipeline.

## 2. Wasserstein Distance Drift Metric

In [2]:
from scipy.stats import wasserstein_distance

# Calculate Wasserstein Distance (Earth Mover's Distance)
w_dist = wasserstein_distance([0, 1, 2], [0, 1, 2], u_weights=actual, v_weights=expected)
print(f"Wasserstein Distance (EMD): {w_dist:.4f}")

assert w_dist > 0.0, "Wasserstein distance check failed!"

Wasserstein Distance (EMD): 0.2000


### Output Analysis: Wasserstein Distance
The computed Wasserstein Distance is `0.2000`. This measures the minimum cost of shifting probability mass to match distributions. Tracking both PSI and Wasserstein Distance provides robust multi-dimensional alerts for data shift detection in production.